pygfx/utils/transform.py

```diff
@@ -0,0 +1,244 @@
+import numpy as np
+import pylinalg as pla
+from datetime import datetime
+
+from typing import List
+
+
+class cached:
+    """Cache for computed properties.
+
+    This descriptor implements a minimal timestamp-based cache for computed
+    properties. The value of the property is computed using ``update_fn`` and
+    the result is cached until ``obj.last_modified`` advances. At this point the
+    value of the computed property is recomputed upon the next read.
+
+
+
+    Example
+    -------
+
+    class Foobar:
+        def __init__(self, x):
+            self._value = 0
+            self.last_modified = datetime.now()
+
+        @property
+        def value(self):
+            return self._value
+
+        @value.setter
+        def value(self, new_value):
+            self._value = new_value
+            self.last_modified = datetime.now()
+
+        @property
+        @cached
+        def double(self):
+            print("Computed value of double..")
+            return 2*x
+
+    """
+
+    def __init__(self, update_fn=None) -> None:
+        self.update_fn = update_fn
+        self.last_updated = None
+        self.cached_value = None
+
+    def __get__(self, obj: "AffineBase", objtype=None):
+        if obj is None:
+            return self
+
+        if self.last_updated is None or obj.last_modified > self.last_updated:
+            self.cached_value = self.update_fn(obj)
+            self.last_updated = obj.last_modified
+
+        return self.cached_value
+
+    def __call__(self, obj):
+        """Helper to allow interoperability with @property."""
+        return self.__get__(obj)
+
+
+class AffineBase:
+    last_modified: datetime
+    matrix: np.ndarray
+
+    @property
+    @cached
+    def inverse_matrix(self):
+        return np.linalg.inv(self.matrix)
+
+    @property
+    @cached
+    def position(self) -> np.ndarray:
+        pos, _, _ = pla.matrix_decompose(self.matrix)
+        return pos
+
+    @property
+    @cached
+    def rotation(self) -> np.ndarray:
+        _, rot, _ = pla.matrix_decompose(self.matrix)
+        return rot
+
+    def __array__(self, dtype=None):
+        return self.matrix.astype(dtype)
+
+    @property
+    @cached
+    def scale(self) -> np.ndarray:
+        _, _, scale = pla.matrix_decompose(self.matrix)
```

Comment:
would it be more efficient to create a single cached prop for scale, rotation and position and then provide three uncached properties that just draw from it? I would think that avoids running matrix_decompose 3x when you're getting all three properties

Reply:
Yes, that's right. I will update.

Reply:
fixed.